src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala

```diff
@@ -0,0 +1,262 @@
+   *    is used on the root of the nested field (ie: `isnotnull(nested#102)`
+   *    does not makes any sense when using the index field).
+   *
+   * For example, given the following query:
+   * {{{
+   *   df
+   *     .filter("nested.leaf.cnt > 10 and nested.leaf.id == 'leaf_id9'")
+   *     .select("Date", "nested.leaf.cnt")
+   * }}}
+   *
+   * Having this simple filter:
+   * {{{
+   *   Filter (isnotnull(nested#102) && (nested#102.leaf.cnt > 10) &&
+   *           (nested#102.leaf.id = leaf_id9))
+   * }}}
+   *
+   * The filter part should become:
+   * {{{
+   *   Filter ((__hs_nested.nested.leaf.cnt#335 > 10) &&
+   *           (__hs_nested.nested#.leaf.id#336 = leaf_id9))
+   * }}}
+   *
+   * @param filter The filter that needs to be transformed.
+   * @return The transformed filter with support for nested indexed fields.
+   */
+  protected[rules] def transformFilter(filter: Filter): Filter = {
+    val names = extractNamesFromExpression(filter.condition)
+    val transformedCondition = filter.condition.transformDown {
+      case bo @ BinaryOperator(IsNotNull(AttributeReference(name, _, _, _)), other) =>
+        if (names.toDiscard.contains(name)) {
+          other
+        } else {
+          bo
+        }
+      case bo @ BinaryOperator(other, IsNotNull(AttributeReference(name, _, _, _))) =>
+        if (names.toDiscard.contains(name)) {
+          other
+        } else {
+          bo
+        }
+      case g: GetStructField =>
+        val n = getChildNameFromStruct(g)
+        if (names.toKeep.contains(n)) {
+          val escapedFieldName = PlanUtils.prefixNestedField(n)
+          getExprId(filter, escapedFieldName) match {
+            case Some(exprId) =>
+              val fieldType = extractTypeFromExpression(g, n)
+              val attr = extractAttributeRef(g, n)
+              attr.copy(escapedFieldName, fieldType, attr.nullable, attr.metadata)(
+                exprId,
+                attr.qualifier)
+            case _ =>
+              g
+          }
+        } else {
+          g
+        }
+      case o =>
+        o
+    }
+    filter.copy(condition = transformedCondition)
+  }
+
+  /**
+   * The method transforms the project part of a plan to support indexes on
+   * nested fields.
+   *
+   * For example, given the following query:
+   * {{{
+   *   df
+   *     .filter("nested.leaf.cnt > 10 and nested.leaf.id == 'leaf_id9'")
+   *     .select("Date", "nested.leaf.cnt")
+   * }}}
+   *
+   * Having this simple projection:
+   * {{{
+   *   Project [Date#100, nested#102.leaf.cnt]
+   * }}}
+   *
+   * The projection part should become:
+   * {{{
+   *   Project [Date#330, __hs_nested.nested.leaf.cnt#335]
+   * }}}
+   *
+   * @param project The project that needs to be transformed.
+   * @return The transformed project with support for nested indexed fields.
+   */
+  protected[rules] def transformProject(project: Project): Project = {
```

Comment:
Does it work for nested fields in expressions? Example: `select my.nested.field1 + my.nested.field2 from mytab`

Reply:
Also, it seems to be a good candidate for a unit test